jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java

```diff
@@ -70,94 +73,57 @@ private Builder() {
      * @param pathInContainer the path in the container file system corresponding to the {@code
      *     sourceFile}
      * @return this
-     * @throws IOException if an exception occurred when recursively listing the directory
+     * @see LayerConfiguration.Builder#addEntry(Path, AbsoluteUnixPath)
      */
-    public Builder addDependencyFile(Path sourceFile, AbsoluteUnixPath pathInContainer)
-        throws IOException {
-      Preconditions.checkNotNull(layerBuilders.get(LayerType.DEPENDENCIES))
-          .addEntryRecursive(sourceFile, pathInContainer);
+    public Builder addFile(LayerType layerType, Path sourceFile, AbsoluteUnixPath pathInContainer) {
+      Preconditions.checkNotNull(layerBuilders.get(layerType))
+          .addEntry(sourceFile, pathInContainer);
       return this;
     }
 
     /**
-     * Adds a file to the snapshot dependency layer. If the source file is a directory, the
-     * directory and its contents will be added recursively. See {@link
-     * LayerConfiguration.Builder#addEntryRecursive} for concrete examples about how the file will
-     * be placed in the image.
+     * Adds directory contents to a layer selectively and recursively. {@code sourceRoot} must be a
+     * directory. Empty directories will always be added regardless of {@code pathFilter}, except
+     * that {@code sourceRoot} is never added.
      *
-     * @param sourceFile the source file to add to the layer
-     * @param pathInContainer the path in the container file system corresponding to the {@code
-     *     sourceFile}
-     * @return this
-     * @throws IOException if an exception occurred when recursively listing the directory
-     */
-    public Builder addSnapshotDependencyFile(Path sourceFile, AbsoluteUnixPath pathInContainer)
-        throws IOException {
-      Preconditions.checkNotNull(layerBuilders.get(LayerType.SNAPSHOT_DEPENDENCIES))
-          .addEntryRecursive(sourceFile, pathInContainer);
-      return this;
-    }
-
-    /**
-     * Adds a file to the resource layer. If the source file is a directory, the directory and its
-     * contents will be added recursively. See {@link LayerConfiguration.Builder#addEntryRecursive}
-     * for concrete examples about how the file will be placed in the image.
+     * <p>The contents of {@code sourceRoot} will be placed into {@code basePathInContainer}. For
+     * example, if {@code sourceRoot} is {@code /usr/home}, {@code /usr/home/passwd} exists locally,
+     * and {@code basePathInContainer} is {@code /etc}, then the image will have {@code
+     * /etc/passwd}.
      *
-     * @param sourceFile the source file to add to the layer
-     * @param pathInContainer the path in the container file system corresponding to the {@code
-     *     sourceFile}
+     * @param layerType the layer to add files into
+     * @param sourceRoot root directory whose contents will be added
+     * @param pathFilter only the files satisfying the filter will be added, unless the files are
```

Comment:
Maybe reword to describe the parameter itself instead of what's affected by it, e.g. `filter that determines which files (not directories) should be added`